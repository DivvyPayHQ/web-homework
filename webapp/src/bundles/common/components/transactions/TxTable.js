import React from 'react'
import { arrayOf, string, bool, number, shape } from 'prop-types'
import { css } from '@emotion/core'

const styles = css`
 .header {
   font-weight: bold;
 }
`

const makeDataTestId = (transactionId, fieldName) => `transaction-${transactionId}-${fieldName}`

export function TxTable ({ data }) {
  return (
    <table css={styles}>
      <tbody>
        <tr className='header'>
          <td >ID</td>
          <td >User ID</td>
          <td >Description</td>
          <td >Merchant ID</td>
          <td >Debit</td>
          <td >Credit</td>
          <td >Amount</td>
        </tr>
        {
          data.map(tx => {
            const { id, user_id: userId, description, merchant_id: merchantId, debit, credit, amount } = tx
            return (
              <tr data-testid={`transaction-${id}`} key={`transaction-${id}`}>
                <td data-testid={makeDataTestId(id, 'id')}>{id}</td>
                <td data-testid={makeDataTestId(id, 'userId')}>{userId}</td>
                <td data-testid={makeDataTestId(id, 'description')}>{description}</td>
                <td data-testid={makeDataTestId(id, 'merchant')}>{merchantId}</td>
                <td data-testid={makeDataTestId(id, 'debit')}>{debit}</td>
                <td data-testid={makeDataTestId(id, 'credit')}>{credit}</td>
                <td data-testid={makeDataTestId(id, 'amount')}>{amount}</td>
              </tr>
            )
          })
        }
      </tbody>
    </table>

  )
}

TxTable.propTypes = {
  data: arrayOf(shape({
    id: string,
    user_id: string,
    description: string,
    merchant_id: string,
    debit: bool,
    credit: bool,
    amount: number
  }))
}
