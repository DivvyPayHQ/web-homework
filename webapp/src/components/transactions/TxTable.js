/* eslint-disable camelcase, react/jsx-sort-props */
import React, { useState, useEffect, useRef } from 'react'
import { arrayOf, string, bool, number, shape } from 'prop-types'
import { useMutation } from '@apollo/client'
import Paper from '@material-ui/core/Paper'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Button from '@material-ui/core/Button'
import IconButton from '@material-ui/core/IconButton'
import Input from '@material-ui/core/Input'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import AddCircleOutlineIcon from '@material-ui/icons/AddCircleOutline'
import ClearIcon from '@material-ui/icons/Clear'
import DeleteIcon from '@material-ui/icons/Delete'
import DoneIcon from '@material-ui/icons/Done'
import EditIcon from '@material-ui/icons/Edit'
import transactionStyles from '../../styles/transactions'
import AddTransaction from '../../gql/addTransaction.gql'
import UpdateTransaction from '../../gql/updateTransaction.gql'

const makeDataTestId = (transactionId, fieldName) => `transaction-${transactionId}-${fieldName}`

const tableHeaderKeys = [
  { id: 'id', label: 'ID', readOnly: true, placeholder: 'Auto Generated' },
  { id: 'user_id', label: 'User ID', placeholder: 'employee1' },
  { id: 'description', label: 'Description', placeholder: 'groceries' },
  { id: 'merchant_id', label: 'Merchant ID', placeholder: 'walmart' },
  { id: 'amount', label: 'Amount', placeholder: '-100 or +100' }
]

const formatter = new Intl.NumberFormat('en-US', {
  style: 'currency',
  currency: 'USD'
})

const createData = (data) => data.map(({ id, user_id, description, merchant_id, debit, amount }) => ({
  id,
  user_id,
  description,
  merchant_id,
  amount,
  debit,
  isEditMode: false
}))

const TxTable = ({ data }) => {
  const [addTransaction] = useMutation(AddTransaction)
  const [updateTransaction] = useMutation(UpdateTransaction)
  const [rows, setRows] = useState(createData(data))
  const [previous, setPrevious] = useState({})
  const classes = transactionStyles()
  let resettingIdRef = useRef(null)

  const onToggleEditMode = id => {
    setRows(state => {
      return rows.map(row => {
        if (row.id === id) {
          return { ...row, isEditMode: !row.isEditMode }
        }
        return row
      })
    })
  }

  const onChange = (e, row) => {
    if (!previous[row.id]) {
      setPrevious(state => ({ ...state, [row.id]: row }))
    }
    const { value, name } = e.target
    const { id } = row
    const newRows = rows.map(row => {
      if (row.id === id) {
        return { ...row, [name]: value }
      }
      return row
    })
    setRows(newRows)
  }

  const onRevert = id => {
    const newRows = rows.map(row => {
      if (row.id === id) {
        return previous[id] ? previous[id] : row
      }
      return row
    })
    console.log(newRows)
    setRows(newRows)
    setPrevious(state => {
      delete state[id]
      return state
    })
    resettingIdRef.current = id
  }

  const onDelete = id => {
    const newRows = rows.filter(row => row.id !== id)
    setRows(newRows)
    setPrevious(state => {
      delete state[id]
      return state
    })
    resettingIdRef.current = id
  }

  const onSave = async id => {
    if (id === '') {
      const newTransaction = rows.find(row => row.id === '')
      newTransaction.credit = /^\+/.test(newTransaction.amount) || false
      newTransaction.debit = /^-/.test(newTransaction.amount) || false
      newTransaction.amount = parseFloat(newTransaction.amount.substring(1))
      delete newTransaction.isEditMode
      delete newTransaction.id
      try {
        const { data } = await addTransaction({
          variables: {
            ...newTransaction
          }
        })

        const newRows = rows.map(row => {
          if (row.id === '') {
            return { ...row, id: data.addTransaction.id, isEditMode: false }
          }
          return row
        })
        setRows(newRows)
      } catch (err) {
        console.log(err)
      }
    } else {
      const updatedTransaction = rows.find(row => row.id === id)
      if (updatedTransaction.amount !== previous[id].amount) {
        updatedTransaction.credit = /^\+/.test(updatedTransaction.amount) || false
        updatedTransaction.debit = /^-/.test(updatedTransaction.amount) || false
        updatedTransaction.amount = parseFloat(updatedTransaction.amount.substring(1))
      }
      delete updatedTransaction.isEditMode
      try {
        await updateTransaction({
          variables: {
            ...updatedTransaction
          }
        })
        setRows(state => {
          return rows.map(row => {
            if (row.id === id) {
              delete row.credit
              return { ...row }
            }
            return row
          })
        })
      } catch (err) {
        console.log(err)
      }
    }
  }

  const onAddTransactionClick = () => {
    const newRows = [...rows]
    const emptyRow = {
      id: '',
      user_id: '',
      description: '',
      merchant_id: '',
      amount: '',
      isEditMode: true
    }
    newRows.unshift(emptyRow)
    setRows(newRows)
  }

  useEffect(() => {
    if (resettingIdRef.current) {
      onToggleEditMode(resettingIdRef.current)
    }
    resettingIdRef.current = false
  }, [rows])

  return (
    <>
      <Grid container justifyContent='space-between' alignContent='center' className={classes.heading}>
        <Typography component='h1' variant='h3'>Transactions</Typography>
        <Button
          variant='contained'
          color='primary'
          onClick={() => onAddTransactionClick()}
          startIcon={<AddCircleOutlineIcon />}
          data-testid={'transaction-add-button'}
        >
          Add Transaction
        </Button>
      </Grid>
      <Paper>
        <TableContainer>
          <Table stickyHeader>
            <TableHead>
              <TableRow>
                <TableCell align='left' />
                {tableHeaderKeys.map(header => (
                  <TableCell align='left' key={header.id}>
                    {header.label}
                  </TableCell>
                ))}
              </TableRow>
            </TableHead>
            <TableBody>
              {rows.map(row => (
                <TableRow key={row.id} data-testid={makeDataTestId(row.id, 'table-row')} >
                  <TableCell className={classes.selectTableCell}>
                    {row.isEditMode ? (
                      <>
                        <IconButton
                          aria-label='save'
                          onClick={() => onSave(row.id)}
                          data-testid={makeDataTestId(row.id, 'save-button')}
                        >
                          <DoneIcon />
                        </IconButton>
                        <IconButton
                          aria-label='revert'
                          onClick={() => onRevert(row.id)}
                          data-testid={makeDataTestId(row.id, 'revert-button')}
                        >
                          <ClearIcon />
                        </IconButton>
                      </>
                    ) : (
                      <>
                        <IconButton
                          aria-label='edit'
                          onClick={() => onToggleEditMode(row.id)}
                          data-testid={makeDataTestId(row.id, 'edit-button')}
                        >
                          <EditIcon />
                        </IconButton>
                        <IconButton
                          aria-label='delete'
                          onClick={() => onDelete(row.id)}
                          data-testid={makeDataTestId(row.id, 'delete-button')}
                        >
                          <DeleteIcon />
                        </IconButton>
                      </>
                    )}
                  </TableCell>
                  {Object.keys(row).map(key => {
                    if (row.id === '60f5c0bc32fb272a902e5d86') { console.log(key) }
                    return key !== 'isEditMode' && key !== 'debit' && (
                      <TableCell align='left' className={classes.tableCell} key={key} data-testid={makeDataTestId(row.id, `${key}-cell`)} >
                        {row.isEditMode && key !== 'id' ? (
                          <Input
                            value={row[key]}
                            name={key}
                            onChange={e => onChange(e, row)}
                            className={classes.input}
                            placeholder={tableHeaderKeys.find(header => header.id === key).placeholder || ''}
                            inputProps={{ readOnly: tableHeaderKeys.find(header => header.id === key).readOnly || false }}
                            disabled={tableHeaderKeys.find(header => header.id === key).readOnly || false}
                            data-testid={makeDataTestId(row.id, `${key}-input`)}
                          />
                        ) : (
                          <>
                            {key === 'amount' ? (
                              <Typography style={{ color: row.debit ? 'red' : 'green' }}>
                                {`${row.debit ? '-' : '+'} ${formatter.format(row[key])}`}
                              </Typography>
                            ) : row[key]}
                          </>
                        )}
                      </TableCell>
                    )
                  })}
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </TableContainer>
      </Paper>
    </>
  )
}

TxTable.propTypes = {
  data: arrayOf(shape({
    id: string,
    user_id: string,
    description: string,
    merchant_id: string,
    debit: bool,
    credit: bool,
    amount: number
  }))
}

export default TxTable
