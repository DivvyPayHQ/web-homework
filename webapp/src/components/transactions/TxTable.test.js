import { fireEvent, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { MockedProvider } from '@apollo/client/testing'
import TxTable from './TxTable'
import { transactions } from '../../../mocks/transactions-data'

describe('Transactions Table', () => {
  const mocks = []
  it('should match snapshot', () => {
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <TxTable data={transactions} />
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
  })
  it('should show user "employee4" with amount "150"', () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypeName={false}>
        <TxTable data={transactions} />
      </MockedProvider>
    )
    const employee4Tx = transactions.find(tx => tx.user_id === 'employee4' && tx.amount === 150)
    const employee4TxRow = getByTestId(`transaction-${employee4Tx.id}-table-row`)
    expect(employee4TxRow).toBeInTheDocument()
  })
  it('should switch to edit mode when edit icon clicked', () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypeName={false}>
        <TxTable data={transactions} />
      </MockedProvider>
    )
    const employee4Tx = transactions.find(tx => tx.user_id === 'employee4' && tx.amount === 150)
    const employee4EditButton = getByTestId(`transaction-${employee4Tx.id}-edit-button`)
    fireEvent.click(employee4EditButton)
    const employee4InputField = getByTestId(`transaction-${employee4Tx.id}-user_id-input`)
    expect(employee4InputField).toBeInTheDocument()
  })
  it('should add transaction row when Add Transaction button clicked', () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypeName={false}>
        <TxTable data={transactions} />
      </MockedProvider>
    )
    const txAddButton = getByTestId(`transaction-add-button`)
    fireEvent.click(txAddButton)
    const newInputField = getByTestId(`transaction--user_id-input`)
    expect(newInputField).toBeInTheDocument()
  })
  it('should delete row when delete button clicked', () => {
    const { getByTestId, queryByTestId } = render(
      <MockedProvider mocks={mocks} addTypeName={false}>
        <TxTable data={transactions} />
      </MockedProvider>
    )
    const employee4Tx = transactions.find(tx => tx.user_id === 'employee4' && tx.amount === 150)
    const employee4DeleteButton = getByTestId(`transaction-${employee4Tx.id}-delete-button`)
    fireEvent.click(employee4DeleteButton)
    const deletedRow = queryByTestId(`transaction--user_id-input`)
    expect(deletedRow).not.toBeInTheDocument()
  })
})
