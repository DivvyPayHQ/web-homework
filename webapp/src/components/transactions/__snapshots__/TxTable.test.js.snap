// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions Table should match snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root makeStyles-heading-1 MuiGrid-container MuiGrid-align-content-xs-center MuiGrid-justify-content-xs-space-between"
  >
    <h1
      class="MuiTypography-root MuiTypography-h3"
    >
      Transactions
    </h1>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      data-testid="transaction-add-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
        </span>
        Add Transaction
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiTableContainer-root"
    >
      <table
        class="MuiTable-root MuiTable-stickyHeader"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            />
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            >
              ID
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            >
              User ID
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            >
              Description
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            >
              Merchant ID
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            >
              Amount
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root"
            data-testid="transaction-5d5c1f747e01cd704f18f863-table-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-selectTableCell-2"
            >
              <button
                aria-label="edit"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transaction-5d5c1f747e01cd704f18f863-edit-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transaction-5d5c1f747e01cd704f18f863-delete-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f863-id-cell"
            >
              5d5c1f747e01cd704f18f863
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f863-user_id-cell"
            >
              employee4
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f863-description-cell"
            >
              cleaningsupplies
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f863-merchant_id-cell"
            >
              walmart
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f863-amount-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="color: red;"
              >
                - $150.00
              </p>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
            data-testid="transaction-5d5c1f747e01cd704f18f864-table-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-selectTableCell-2"
            >
              <button
                aria-label="edit"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transaction-5d5c1f747e01cd704f18f864-edit-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transaction-5d5c1f747e01cd704f18f864-delete-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f864-id-cell"
            >
              5d5c1f747e01cd704f18f864
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f864-user_id-cell"
            >
              employee3
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f864-description-cell"
            >
              refund
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f864-merchant_id-cell"
            >
              walmart
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f864-amount-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="color: green;"
              >
                + $250.00
              </p>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
            data-testid="transaction-5d5c1f747e01cd704f18f865-table-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-selectTableCell-2"
            >
              <button
                aria-label="edit"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transaction-5d5c1f747e01cd704f18f865-edit-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transaction-5d5c1f747e01cd704f18f865-delete-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f865-id-cell"
            >
              5d5c1f747e01cd704f18f865
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f865-user_id-cell"
            >
              employee5
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f865-description-cell"
            >
              refund
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f865-merchant_id-cell"
            >
              walmart
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transaction-5d5c1f747e01cd704f18f865-amount-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="color: green;"
              >
                + $100.00
              </p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
